src/JuliusSweetland.OptiKey.Core/Services/KeyboardOutputService.cs

```diff
@@ -1119,7 +1142,14 @@ private void PublishKeyPress(char character)
                         releaseAlt = true;
                     }
 
-                    publishService.KeyDownUp((VirtualKeyCode)vkCode);
+                    if (type == KeyPressKeyValue.KeyPressType.Press)
```

Comment:
Looks good, but it is worth considering what happens around the modifier keys. The way this logic works is a query is made to the operating system to figure out which key strokes achieve the character we want to output, based on the user's keyboard layout. This might be that we need to simulate the ALT key being down to be able to fire the correct keystrokes to output the character. This is fine if our action is to press the key down and up as the sequence becomes:
ALT DOWN
KEY DOWN
KEY UP
ALT UP

But what happens if this keypress is part of a dynamic keyboard and the key is only toggled. That would result in:

Dynamic key press 1:
ALT DOWN
KEY DOWN
ALT UP

Dynamic key press 2:
ALT DOWN
KEY UP
ALT UP

Is that going to cause unexpected behaviour? I don't know how the OS, or other apps, might interpret the release of the ALT key - would they think the key press sequence is complete, or might they wait for the release of the actual key...I don't know. Just flagging as a potential pain point. Might be worth adding a few notes in a comment here so we can consider this again if an issue does arise in future.

Reply:
I hear you. I've been internally debating what makes the most sense for KeyDown/Up/Toggle in regard to obeying other rules and interacting with modifiers. I'm leaning toward KeyDown/Up/Toggle being considered special use case not necessarily expected to conform to other rules.

Reply:
@AdamRoden I think that's fair, but it may be worth adding extra WARN logging in situations where a modifier key(s) would need to be pressed to achieve the virtual key code - we can continue doing that, but log that the behaviour may be different from what is expected as the modifier will be auto released after the KeyUp/Down/Toggle event has been fired.